### PR TITLE
Add support for `txt` file extensions and display hash for non text/png files

### LINKED ISSUE #406: Difference viewer reports "File without changes" although they are reported as changed in console
On the shell:
```
user@host:shiny$ cat app.R

shiny::shinyApp(
    shinyUI(fluidPage("Hello", shinyWidgets::dropdownButton("World!"))),
    function(input, output) {}
)

user@host:shiny$ cat tests/shinytest/test.R 

app <- ShinyDriver$new("../../")
app$snapshotInit("test")
app$snapshot(filename = "test")
```

In R:
```
> shinytest::testApp()
Loading required package: shiny
Running test.R 
==== Comparing test... 
  No existing snapshot at './tests/shinytest/test-expected/' This is a first run of tests.

Updating baseline results at ./tests/shinytest/test-expected...
Renaming tests/shinytest/test-current
      => ./tests/shinytest/test-expected.
> shinytest::testApp()
Running test.R 
==== Comparing test... 
  Differences detected between test-current/ and test-expected/:

    Name         Status      
    test      != Files differ
    test.png     No change   
Would you like to view the differences between expected and current results [y/n]? y
Differences in current results found for: 
* test
Viewing diff for test

Listening on http://127.0.0.1:1234
```

image

In shell:
```
user@host:shiny$ diff tests/shinytest/test-current/test tests/shinytest/test-expected/test
3c3
<     "drop158653462": 0
---
>     "drop11000393": 0
```

TLDR: Text files are significantly different (not just line endings). `testApp()` reports them as different. `viewTestDiff()` does not.

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #406 

* Allow for diffs of `*.txt` files
* Execute `digest::digest(raw_data, algo = "sha1")` on non `.(json|download|txt|png)` files
* Display a message containing the hash and which files are supported.
* Add docs to `app$snapshot` to encourage using a `.json` file extension

Before PR:
image

With PR:

Screen Shot 2021-09-09 at 4 01 37 PM

---------------------

SHA1 spelling came from https://en.wikipedia.org/wiki/SHA-1

Benchmark of `sha1`, `md5`, and `spookyhash`.

```
a <- read_raw("pr407/tests/shinytest/test-expected-old/test")

# "small" file
bench::mark(
  md5 = digest::digest(a),
  spooky = digest::digest(a, algo = "spookyhash"),
  sha1 = digest::digest(a, algo = "sha1"),
  check = FALSE
)
#> # A tibble: 3 × 13
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory     time      gc
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list>     <list>    <list>
#> 1 md5          18.8µs   25.5µs    38118.    8.63KB    11.4   9997     3      262ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 2 spooky         24µs   30.7µs    32743.    8.63KB     9.83  9997     3      305ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 3 sha1         21.7µs   28.9µs    35311.    8.63KB    14.1   9996     4      283ms <NULL> <Rprofmem… <bench_t… <tibble …

# 1000x sizeof("small" file)
bench::mark(
  md5 = digest::digest(b),
  spooky = digest::digest(b, algo = "spookyhash"),
  sha1 = digest::digest(b, algo = "sha1"),
  check = FALSE
)
#> # A tibble: 3 × 13
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory     time      gc
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list>     <list>    <list>
#> 1 md5         212.8µs  262.2µs     3400.  101.19KB     6.43  1585     3      466ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 2 spooky       30.1µs   36.1µs    25985.    8.63KB     7.80  9997     3      385ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 3 sha1        216.7µs  236.6µs     3981.   90.73KB     8.63  1846     4      464ms <NULL> <Rprofmem… <bench_t… <tibble …
```

The results are still less than a millisecond, so using the familiar `sha1` as the result does not need to be crypto perfect, just signal differences.

### CODE REVIEW

File: R/view-diff.R [hunk 1]
```diff
@@ -41,12 +41,17 @@ diffviewer_widget <- function(old, new, width = NULL, height = NULL,
...
-    if (grepl("\\.json$", filename) || grepl("\\.download$", filename)) {
+    if (grepl("\\.(json|download|txt)$", filename)) {
       raw_to_utf8(bin_data)
     } else if (grepl("\\.png$", filename)) {
       paste0("data:image/png;base64,", jsonlite::base64_enc(bin_data))
     } else {
-      ""
+      # provide hash of file contents as a proxy to display binary differences
+      paste0(
+        "{shinytest} - [Spooky]Hash of file contents: ", digest::digest(bin_data, algo = "spookyhash"), "\n",
```
Comment (wch) [inline R/view-diff.R:51]: How about using a more "standard" hash algorithm, like SHA1?

Comment (schloerke) [reply R/view-diff.R:51]: Using the same setup as above, `sha1` is as slow as `md5`. Where as `spookyhash` is faster than `sha1`/`md5` for larger files.

```r
# "small" file
bench::mark(
  md5 = digest::digest(a),
  spooky = digest::digest(a, algo = "spookyhash"),
  sha1 = digest::digest(a, algo = "sha1"),
  check = FALSE
)
#> # A tibble: 3 × 13
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory     time      gc
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list>     <list>    <list>
#> 1 md5          18.8µs   25.5µs    38118.    8.63KB    11.4   9997     3      262ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 2 spooky         24µs   30.7µs    32743.    8.63KB     9.83  9997     3      305ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 3 sha1         21.7µs   28.9µs    35311.    8.63KB    14.1   9996     4      283ms <NULL> <Rprofmem… <bench_t… <tibble …
```
```r
# 1000x sizeof("small" file)
bench::mark(
  md5 = digest::digest(b),
  spooky = digest::digest(b, algo = "spookyhash"),
  sha1 = digest::digest(b, algo = "sha1"),
  check = FALSE
)
#> # A tibble: 3 × 13
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory     time      gc
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list>     <list>    <list>
#> 1 md5         212.8µs  262.2µs     3400.  101.19KB     6.43  1585     3      466ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 2 spooky       30.1µs   36.1µs    25985.    8.63KB     7.80  9997     3      385ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 3 sha1        216.7µs  236.6µs     3981.   90.73KB     8.63  1846     4      464ms <NULL> <Rprofmem… <bench_t… <tibble …
```

-------------------

I'm ok with using sha1 as it is familiar to users.

Comment (schloerke) [reply R/view-diff.R:51]: Ah. It is micro seconds... Going to switch to `sha1`.

**Comment by schloerke** [comment]
All comments addressed. Merging.

**Pull request merged**